<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Searching protein sequence</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p>This vignette assumes that you have already properly installed rMAUPS.</p>

<h1>Searching protein sequence</h1>

<p>First, load the rMAUPS package.</p>

<pre><code class="r">library(rMAUPS)
</code></pre>

<p>Next, read in protein sequences from uniprot.</p>

<pre><code class="r">uniprot_path &lt;- system.file(&quot;extdata&quot;, &quot;human_uniprot_seq.txt&quot;, package = &quot;rMAUPS&quot;)
uniprot &lt;- read.delim(uniprot_path, sep=&#39;\t&#39;)
head(uniprot)
</code></pre>

<pre><code>##      gene     ID
## 1    KRAS P01116
## 2 CYP4F11 Q9HBI6
## 3    CDH5 P33151
## 4   CCDC8 Q9H0W5
## 5   CEP76 Q8TAP6
## 6   CEP72 Q9P209
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   protein_sequence
## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPTIEDSYRKQVVIDGETCLLDILDTAGQEEYSAMRDQYMRTGEGFLCVFAINNTKSFEDIHHYREQIKRVKDSEDVPMVLVGNKCDLPSRTVDTKQAQDLARSYGIPFIETSAKTRQRVEDAFYTLVREIRQYRLKKISKEEKTPGCVKIKKCIIM
## 2                                                                                                                                                                                                                                                                     MPQLSLSWLGLGPVAASPWLLLLLVGGSWLLARVLAWTYTFYDNCRRLQCFPQPPKQNWFWGHQGLVTPTEEGMKTLTQLVTTYPQGFKLWLGPTFPLLILCHPDIIRPITSASAAVAPKDMIFYGFLKPWLGDGLLLSGGDKWSRHRRMLTPAFHFNILKPYMKIFNKSVNIMHDKWQRLASEGSARLDMFEHISLMTLDSLQKCVFSFESNCQEKPSEYIAAILELSAFVEKRNQQILLHTDFLYYLTPDGQRFRRACHLVHDFTDAVIQERRCTLPTQGIDDFLKNKAKSKTLDFIDVLLLSKDEDGKELSDEDIRAEADTFMFEGHDTTASGLSWVLYHLAKHPEYQEQCRQEVQELLKDREPIEIEWDDLAQLPFLTMCIKESLRLHPPVPVISRCCTQDFVLPDGRVIPKGIVCLINIIGIHYNPTVWPDPEVYDPFRFDQENIKERSPLAFIPFSAGPRNCIGQAFAMAEMKVVLALTLLHFRILPTHTEPRRKPELILRAEGGLWLRVEPLGANSQ
## 3 MQRLMMLLATSGACLGLLAVAAVAAAGANPAQRDTHSLLPTHRRQKRDWIWNQMHIDEEKNTSLPHHVGKIKSSVSRKNAKYLLKGEYVGKVFRVDAETGDVFAIERLDRENISEYHLTAVIVDKDTGENLETPSSFTIKVHDVNDNWPVFTHRLFNASVPESSAVGTSVISVTAVDADDPTVGDHASVMYQILKGKEYFAIDNSGRIITITKSLDREKQARYEIVVEARDAQGLRGDSGTATVLVTLQDINDNFPFFTQTKYTFVVPEDTRVGTSVGSLFVEDPDEPQNRMTKYSILRGDYQDAFTIETNPAHNEGIIKPMKPLDYEYIQQYSFIVEATDPTIDLRYMSPPAGNRAQVIINITDVDEPPIFQQPFYHFQLKENQKKPLIGTVLAMDPDAARHSIGYSIRRTSDKGQFFRVTKKGDIYNEKELDREVYPWYNLTVEAKELDSTGTPTGKESIVQVHIEVLDENDNAPEFAKPYQPKVCENAVHGQLVLQISAIDKDITPRNVKFKFILNTENNFTLTDNHDNTANITVKYGQFDREHTKVHFLPVVISDNGMPSRTGTSTLTVAVCKCNEQGEFTFCEDMAAQVGVSIQAVVAILLCILTITVITLLIFLRRRLRKQARAHGKSVPEIHEQLVTYDEEGGGEMDTTSYDVSVLNSVRRGGAKPPRPALDARPSLYAQVQKPPRHAPGAHGGPGEMAAMIEVKKDEADHDGDGPPYDTLHIYGYEGSESIAESLSSLGTDSSDSDVDYDFLNDWGPRFKMLAELYGSDPREELLY
## 4                                                                                                                                                                                                                                                       MLQIGEDVDYLLIPREVRLAGGVWRVISKPATKEAEFRERLTQFLEEEGRTLEDVARIMEKSTPHPPQPPKKPKEPRVRRRVQQMVTPPPRLVVGTYDSSNASDSEFSDFETSRDKSRQGPRRGKKVRKMPVSYLGSKFLGSDLESEDDEELVEAFLRRQEKQPSAPPARRRVNLPVPMFEDNLGPQLSKADRWREYVSQVSWGKLKRRVKGWAPRAGPGVGEARLASTAVESAGVSSAPEGTSPGDRLGNAGDVCVPQASPRRWRPKINWASFRRRRKEQTAPTGQGADIEADQGGEAADSQREEAIADQREGAAGNQRAGAPADQGAEAADNQREEAADNQRAGAPAEEGAEAADNQREEAADNQRAEAPADQRSQGTDNHREEAADNQRAEAPADQGSEVTDNQREEAVHDQRERAPAVQGADNQRAQARAGQRAEAAHNQRAGAPGIQEAEVSAAQGTTGTAPGARARKQVKTVRFQTPGRFSWFCKRRRAFWHTPRLPTLPKRVPRAGEARNLRVLRAEARAEAEQGEQEDQL
## 5                                                                                                                              MSLPPEKASELKQLIHQQLSKMDVHGRIREILAETIREELAPDQQHLSTEDLIKALRRRGIIDDVMKELNFVTDSVEQELPSSPKQPICFDRQSTLKKTNIDPTRRYLYLQVLGGKAFLEHLQEPEPLPGQVCSTFTLCLHYRNQRFRSKPVPCACEPDFHDGFLLEVHRESLGDGTRMADSTTMLSISDPIHMVLIKTDIFGETTLVASYFLEWRSVLGSENGVTSLTVELMGVGTESKVSVGILNIKLEMYPPLNQTLSQEVVNTQLALERQKTAEKERLFLVYAKQWWREYLQIRPSHNSRLVKIFAQDENGINRPVCSYVKPLRAGRLLDTPRQAARFVNVLGYERAPVIGGGGKQEQWCTLLAFLCRNKGDCEDHANLLCSLLLGYGLEAFVCVGTKAKGVPHAWVMTCGTDGAITFWESLTGHRYIHKPTNPDEPPVAEQPKPLYPYRTIGCVFNHQMFLGNCQPSDAVETCVFDLNDESKWKPMSEEAIKSVCAPGATTSLPPFPPLCASTIDASVTSNEIEMQLRLLVSEHRKDLGLTTVWEDQLSYLLSPALASYEFERTTSISAGNEEFQDAIRRAVPDGHTFKGFPIHFVYRNARRAFATCLRSPFCEEIICCRGDQVRLAVRVRVFTYPESACAVWIMFACKYRSVL
## 6                                                                                                                                          MARAGPRLVLSEEAVRAKSGLGPHRDLAELQSLSIPGTYQEKITHLGHSLMSLTGLKSLDLSRNSLVSLEGIQYLTALESLNLYYNCISSLAEVFRLHALTELVDVDFRLNPVVKVEPDYRLFVVHLLPKLQQLDDRPVRASERKASRLHFASEDSLDSKESVPASLKEGRPHHPRAKCTEALAKQSLVMDADDEAVLNLIAECEWDLGRPPGSTSFSQKGREADSRGSQESRHLLSPQLVQYQCGDSGKQGRETRRSSCRGCCLEKMPWSQLCGELPPLYGAEPEASRAPRPHTYFTPHPDSMDTEDSASSQKLDLSGEMVPGPLPAPGKCRKRRMPVGRFQTFSDQEGLGCPERTHGSSVPKESLSRQDSSESRNGRTLSQPEASETEEQRSRGVTDTREPSPGSHSALPGKKTALQAALLETLLDLVDRSWGGCRSLHSNEAFLAQARHILSSVEEFTAAQDSSAMVGEDVGSLALESKSLQSRLAEQQQQHAREMSEVTAELHHTHKELDDLRQHLDKSLEENSRLKSLLLSMKKEVKSADTAATLNLQIAGLQTSVKRLCGEIVELKQHLEHYDKIQELTQMLQESHSSLVSTNEHLLQELSQVRAQHRAEVEQMHWSYQELKKTMALFPHSSASHGGCQAC
</code></pre>

<p>Next, search the protein sequences for lysine residues (&ldquo;K&rdquo;). </p>

<pre><code class="r">lysine_pos &lt;- searchProtSeq(head(uniprot), &#39;K&#39;)
head(lysine_pos)
</code></pre>

<pre><code>##   UniprotId start end
## 1    P01116     5   5
## 2    P01116    16  16
## 3    P01116    42  42
## 4    P01116    88  88
## 5    P01116   101 101
## 6    P01116   104 104
</code></pre>

<p>Note, for the sake of a quick example, the above command only searched against the top few protein sequences. Additionally, although not used here, the searchProtSeq function can also handle searching for sequences using regular expression.</p>

<h1>Viewing protein structure</h1>

<p>First, we&#39;ll view all of the lysine residues for KRAS.</p>

<pre><code class="r"># get hits for P01116
id &lt;- &#39;P01116&#39;
start &lt;- lysine_pos[lysine_pos[&#39;UniprotId&#39;]==id, &#39;start&#39;]
end &lt;- lysine_pos[lysine_pos[&#39;UniprotId&#39;]==id, &#39;end&#39;]

# view on protein structure
browseProtStructure(id, start, end)
</code></pre>

<pre><code>## https://mupit.icm.jhu.edu/MuPIT_Interactive/?gm=P01116:5,P01116:16,P01116:42,P01116:88,P01116:101,P01116:104,P01116:117,P01116:128,P01116:147,P01116:169,P01116:170,P01116:173,P01116:176,P01116:182,P01116:184,P01116:185&amp;protquery=y
</code></pre>

<p>Next, we&#39;ll compare those to the lysines with reported ubiquitination sites.</p>

<pre><code class="r"># Here, we manually create a dataframe for the KRAS ubiqutination sites.
# In the realisitic scenario, you should load the &quot;Ubiquitination_site_dataset&quot; file
# from PhosphositePlus DB using the readPSPUbiquitin function. You&#39;ll need to download
# it your self due to copy protection.
id &lt;- &#39;P01116&#39;
ub &lt;- data.frame(gene=&#39;KRAS&#39;, ID=id, position=c(117, 128, 147))

# get ub sites for P01116
start &lt;- ub[ub[&#39;ID&#39;]==id, &#39;position&#39;]
end &lt;- ub[ub[&#39;ID&#39;]==id, &#39;position&#39;]

# view on protein structure
browseProtStructure(id, start, end)
</code></pre>

<pre><code>## https://mupit.icm.jhu.edu/MuPIT_Interactive/?gm=P01116:117,P01116:128,P01116:147&amp;protquery=y
</code></pre>

<h1>Session info</h1>

<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS  10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] C/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] rMAUPS_0.1.1  ggplot2_3.3.2
## 
## loaded via a namespace (and not attached):
##   [1] readxl_1.3.1                backports_1.1.10           
##   [3] BiocFileCache_1.12.1        sn_1.6-2                   
##   [5] plyr_1.8.6                  GSEABase_1.50.1            
##   [7] splines_4.0.2               BiocParallel_1.22.0        
##   [9] TH.data_1.0-10              GenomeInfoDb_1.24.2        
##  [11] digest_0.6.26               foreach_1.5.1              
##  [13] htmltools_0.5.0             magrittr_1.5               
##  [15] memoise_1.1.0               doParallel_1.0.16          
##  [17] openxlsx_4.2.2              limma_3.44.3               
##  [19] annotate_1.66.0             msmsTests_1.26.0           
##  [21] matrixStats_0.57.0          sandwich_3.0-0             
##  [23] askpass_1.1                 prettyunits_1.1.1          
##  [25] colorspace_1.4-1            msmsEDA_1.26.0             
##  [27] blob_1.2.1                  rappdirs_0.3.1             
##  [29] ggrepel_0.8.2               haven_2.3.1                
##  [31] rbibutils_1.3               xfun_0.18                  
##  [33] dplyr_1.0.2                 jsonlite_1.7.2             
##  [35] crayon_1.3.4                RCurl_1.98-1.2             
##  [37] graph_1.66.0                genefilter_1.70.0          
##  [39] impute_1.62.0               zoo_1.8-8                  
##  [41] survival_3.2-7              iterators_1.0.13           
##  [43] glue_1.4.2                  gtable_0.3.0               
##  [45] zlibbioc_1.34.0             XVector_0.28.0             
##  [47] DelayedArray_0.14.1         car_3.0-10                 
##  [49] BiocGenerics_0.34.0         msigdbr_7.2.1              
##  [51] abind_1.4-5                 scales_1.1.1               
##  [53] vsn_3.56.0                  mvtnorm_1.1-1              
##  [55] DBI_1.1.0                   edgeR_3.30.3               
##  [57] rstatix_0.6.0               Rcpp_1.0.5                 
##  [59] plotrix_3.7-8               metap_1.4                  
##  [61] mzR_2.22.0                  xtable_1.8-4               
##  [63] progress_1.2.2              tmvnsim_1.0-2              
##  [65] foreign_0.8-80              bit_4.0.4                  
##  [67] preprocessCore_1.50.0       stats4_4.0.2               
##  [69] GSVA_1.36.3                 httr_1.4.2                 
##  [71] gplots_3.1.0                RColorBrewer_1.1-2         
##  [73] TFisher_0.2.0               ellipsis_0.3.1             
##  [75] farver_2.0.3                pkgconfig_2.0.3            
##  [77] XML_3.99-0.5                dbplyr_1.4.4               
##  [79] locfit_1.5-9.4              labeling_0.4.2             
##  [81] reshape2_1.4.4              tidyselect_1.1.0           
##  [83] rlang_0.4.8                 later_1.1.0.1              
##  [85] AnnotationDbi_1.50.3        munsell_0.5.0              
##  [87] cellranger_1.1.0            tools_4.0.2                
##  [89] generics_0.1.0              RSQLite_2.2.1              
##  [91] mathjaxr_1.0-1              broom_0.7.2                
##  [93] evaluate_0.14               stringr_1.4.0              
##  [95] fastmap_1.0.1               mzID_1.26.0                
##  [97] knitr_1.30                  bit64_4.0.5                
##  [99] zip_2.1.1                   caTools_1.18.0             
## [101] purrr_0.3.4                 ncdf4_1.17                 
## [103] mime_0.9                    xml2_1.3.2                 
## [105] biomaRt_2.44.4              compiler_4.0.2             
## [107] shinythemes_1.1.2           curl_4.3                   
## [109] affyio_1.58.0               ggsignif_0.6.0             
## [111] tibble_3.0.4                geneplotter_1.66.0         
## [113] stringi_1.5.3               highr_0.8                  
## [115] forcats_0.5.0               MSnbase_2.14.2             
## [117] lattice_0.20-41             ProtGenerics_1.20.0        
## [119] Matrix_1.2-18               markdown_1.1               
## [121] multtest_2.44.0             vctrs_0.3.4                
## [123] mutoss_0.1-12               pillar_1.4.6               
## [125] lifecycle_0.2.0             BiocManager_1.30.10        
## [127] Rdpack_2.0                  MALDIquant_1.19.3          
## [129] data.table_1.13.0           bitops_1.0-6               
## [131] gbRd_0.4-11                 qvalue_2.20.0              
## [133] httpuv_1.5.4                GenomicRanges_1.40.0       
## [135] R6_2.4.1                    pcaMethods_1.80.0          
## [137] affy_1.66.0                 promises_1.1.1             
## [139] KernSmooth_2.23-17          rio_0.5.16                 
## [141] IRanges_2.22.2              codetools_0.2-16           
## [143] gtools_3.8.2                MASS_7.3-53                
## [145] assertthat_0.2.1            SummarizedExperiment_1.18.2
## [147] openssl_1.4.3               DESeq2_1.28.1              
## [149] withr_2.3.0                 mnormt_2.0.2               
## [151] multcomp_1.4-14             S4Vectors_0.26.1           
## [153] GenomeInfoDbData_1.2.3      parallel_4.0.2             
## [155] hms_0.5.3                   grid_4.0.2                 
## [157] tidyr_1.1.2                 carData_3.0-4              
## [159] ggpubr_0.4.0                numDeriv_2016.8-1.1        
## [161] Biobase_2.48.0              shiny_1.5.0
</code></pre>

</body>

</html>
